cache/src/main/java/io/envoyproxy/controlplane/cache/Resources.java

```diff
@@ -30,33 +37,108 @@
 
 public class Resources {
 
+  /**
+   * Version-agnostic representation of a resource. This is useful when the version qualifier
+   * isn't needed.
+   */
+  public enum ResourceType {
+    CLUSTER,
+    ENDPOINT,
+    LISTENER,
+    ROUTE,
+    SECRET
+  }
+
+  public enum ApiVersion {
+    V2,
+    V3
+  }
+
   private static final Logger LOGGER = LoggerFactory.getLogger(Resources.class);
 
   static final String FILTER_ENVOY_ROUTER = "envoy.router";
   static final String FILTER_HTTP_CONNECTION_MANAGER = "envoy.http_connection_manager";
 
-  private static final String TYPE_URL_PREFIX = "type.googleapis.com/envoy.api.v2.";
-
-  public static final String CLUSTER_TYPE_URL = TYPE_URL_PREFIX + "Cluster";
-  public static final String ENDPOINT_TYPE_URL = TYPE_URL_PREFIX + "ClusterLoadAssignment";
-  public static final String LISTENER_TYPE_URL = TYPE_URL_PREFIX + "Listener";
-  public static final String ROUTE_TYPE_URL = TYPE_URL_PREFIX + "RouteConfiguration";
-  public static final String SECRET_TYPE_URL = TYPE_URL_PREFIX + "auth.Secret";
-
-  public static final List<String> TYPE_URLS = ImmutableList.of(
+  public static final String V3_CLUSTER_TYPE_URL = "type.googleapis.com/envoy.config.cluster.v3"
+      + ".Cluster";
+  public static final String V3_ENDPOINT_TYPE_URL = "type.googleapis.com/envoy.config.endpoint.v3"
+      + ".ClusterLoadAssignment";
+  public static final String V3_LISTENER_TYPE_URL = "type.googleapis.com/envoy.config.listener.v3"
+      + ".Listener";
+  public static final String V3_ROUTE_TYPE_URL = "type.googleapis.com/envoy.config.route.v3"
+      + ".RouteConfiguration";
+  public static final String V3_SECRET_TYPE_URL = "type.googleapis.com/envoy.extensions"
+      + ".transport_sockets.tls.v3.Secret";
+
+  private static final String V2_TYPE_URL_PREFIX = "type.googleapis.com/envoy.api.v2.";
+  public static final String CLUSTER_TYPE_URL = V2_TYPE_URL_PREFIX + "Cluster";
```

Comment:
We have `CLUSTER_TYPE_URL` and `V3_ CLUSTER_TYPE_URL`, not `V2_CLUSTER_TYPE_URL` and `V3_ CLUSTER_TYPE_URL`. Why is that?

Also, we could consider introducing a namespace for constants like:
```java
public class Resources {
    // ...
    public static class V2 {
        public static final String CLUSTER_TYPE_URL = "...";
        // ...
    }

    public static class V3 {
        public static final String CLUSTER_TYPE_URL = "...";
        // ...
    }
}
```